csv-extractor/csv_extractor.py

```diff
@@ -131,20 +150,29 @@ def process_csv_file(client, monitor, csv_path, existing_time_series):
             unique_external_ids.add(external_id)
 
     if current_time_series:
-        _log_error(client.datapoints.post_multi_time_series_datapoints, current_time_series)
+        network_threads.append(threading.Thread(target=_log_error, args=(client.datapoints.post_multi_time_series_datapoints, current_time_series)))
 
-    return count_of_data_points, len(unique_external_ids)
+    return count_of_data_points, len(unique_external_ids), network_threads
 
 
 def process_files(client, monitor, paths, time_series_cache, failed_path) -> None:
     """Process one csv file at a time, and either delete it or possibly move it when done."""
     remaining_files_count = len(paths)
     monitor.successfully_processed_files_gauge.set(0)
-
+    network_threads = []
     for path in paths:
         try:
             try:
-                data_points_count, external_ids_count = process_csv_file(client, monitor, path, time_series_cache)
+                count_data_points, count_ext_ids, threads = process_csv_file(client, monitor, path, time_series_cache)
+                network_threads += threads
+                # post the requests if more than 20 or if we're on the last file
```

Comment:
The current logic is that we delete the file after everything was posted successfully. That way we guarantee we don't lose data if there was an error. If error we move the file if that option is enabled.

This new logic will delete files before we send data, and therefor will have no way of preserving it. Sending data after each request has the drawback that we can't parallelize as well as we wish. What do you think?

Reply:
You're right. This is not ideal.

Another possible solution is we parse a file, and post its datapoints parallely, and then move on to the next one.

Which is slightly slower (0.71 seconds/file vs 0.8 seconds/file)

Reply:
I think we should go for the safer logic even if it is slightly slower for now.

Reply:
Please take a look at the new logic